overlord/snapstate/autorefresh.go

```diff
@@ -0,0 +1,248 @@
+)
+
+// the default refresh pattern
+const defaultRefreshSchedule = "00:00-04:59/5:00-10:59/11:00-16:59/17:00-23:59"
+
+// hooks setup by devicestate
+var (
+	CanAutoRefresh         func(st *state.State) (bool, error)
+	RefreshScheduleManaged func(st *state.State) bool
+)
+
+// refreshRetryDelay specified the minimum time to retry failed refreshes
+var refreshRetryDelay = 10 * time.Minute
+
+func resetRefreshScheduleToDefault(st *state.State) (ts []*timeutil.Schedule, scheduleStr string) {
+	refreshSchedule, err := timeutil.ParseSchedule(defaultRefreshSchedule)
+	if err != nil {
+		panic(fmt.Sprintf("defaultRefreshSchedule cannot be parsed: %s", err))
+	}
+	tr := config.NewTransaction(st)
+	tr.Set("core", "refresh.schedule", defaultRefreshSchedule)
+	tr.Commit()
+
+	return refreshSchedule, defaultRefreshSchedule
+}
+
+func autoRefreshInFlight(st *state.State) bool {
+	for _, chg := range st.Changes() {
+		if chg.Kind() == "auto-refresh" && !chg.Status().Ready() {
+			return true
+		}
+	}
+	return false
+}
+
+// autoRefresh will ensure that snaps are refreshed automatically
+// according to the refresh schedule.
+type autoRefresh struct {
+	state *state.State
+
+	lastRefreshSchedule string
+	nextRefresh         time.Time
+	lastRefreshAttempt  time.Time
+}
+
+func newAutoRefresh(st *state.State) *autoRefresh {
+	return &autoRefresh{
+		state: st,
+	}
+}
+
+// RefreshSchedule will return a user visible string with the current
+// schedule for the automatic refreshes.
+func (m *autoRefresh) RefreshSchedule() (string, error) {
+	_, scheduleStr, err := m.refreshScheduleWithDefaultsFallback()
+	return scheduleStr, err
+}
+
+// NextRefresh returns when the next automatic refresh will happen.
+func (m *autoRefresh) NextRefresh() time.Time {
+	return m.nextRefresh
+}
+
+// LastRefresh returns when the last refresh happened.
+func (m *autoRefresh) LastRefresh() (time.Time, error) {
+	var lastRefresh time.Time
+	err := m.state.Get("last-refresh", &lastRefresh)
+	if err != nil && err != state.ErrNoState {
+		return time.Time{}, err
+	}
+	return lastRefresh, nil
+}
+
+// Ensure ensures that we refresh all installed snaps periodically
+func (m *autoRefresh) Ensure() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	// see if it even makes sense to try to refresh
+	if CanAutoRefresh == nil {
+		return nil
+	}
+	if ok, err := CanAutoRefresh(m.state); err != nil || !ok {
+		return err
+	}
+
+	// Check that we have reasonable delays between attempts.
+	// If the store is under stress we need to make sure we do not
+	// hammer it too often
+	if !m.lastRefreshAttempt.IsZero() && m.lastRefreshAttempt.Add(refreshRetryDelay).After(time.Now()) {
+		return nil
+	}
+
+	// get lastRefresh and schedule
+	lastRefresh, err := m.LastRefresh()
+	if err != nil {
+		return err
+	}
+
+	refreshSchedule, refreshScheduleStr, err := m.refreshScheduleWithDefaultsFallback()
+	if err != nil {
+		return err
+	}
+	if len(refreshSchedule) == 0 {
+		return nil
```

Comment:
should we set m.nextRefresh = time.Time{}, there's no next refresh at this point